IpmiFeaturePkg/GenericIpmi/Common/IpmiInitialize.c

```diff
@@ -0,0 +1,409 @@
+            StatusCodeValue[*ErrorCount] = EFI_COMPUTING_UNIT_FIRMWARE_PROCESSOR | CU_FP_EC_SDR_EMPTY;
+            (*ErrorCount)++;
+          }
+        }
+
+        break;
+
+      case IPMI_APP_SELFTEST_FATAL_HW_ERROR:
+        IpmiInstance->BmcStatus = BMC_HARDFAIL;
+        break;
+
+      default:
+        //
+        // Call routine to check device specific failures.
+        //
+        GetDeviceSpecificTestResults (SelfTestResult, &IpmiInstance->BmcStatus);
+    }
+
+    if (IpmiInstance->BmcStatus == BMC_HARDFAIL) {
+      if (*ErrorCount < MAX_SOFT_COUNT) {
+        StatusCodeValue[*ErrorCount] = EFI_COMPUTING_UNIT_FIRMWARE_PROCESSOR | EFI_CU_FP_EC_HARD_FAIL;
+        (*ErrorCount)++;
+      }
+    } else if (IpmiInstance->BmcStatus == BMC_SOFTFAIL) {
+      if (*ErrorCount < MAX_SOFT_COUNT) {
+        StatusCodeValue[*ErrorCount] = EFI_COMPUTING_UNIT_FIRMWARE_PROCESSOR | EFI_CU_FP_EC_SOFT_FAIL;
+        (*ErrorCount)++;
+      }
+    }
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Execute the Get Device ID command to determine whether or not the BMC is in
+  Force Update Mode.  If it is, then report it to the error manager.
+
+  @param[in]      IpmiInstance      Data structure describing BMC variables and
+                                    used for sending commands
+  @param[in]      StatusCodeValue   An array used to accumulate error codes for
+                                    later reporting.
+  @param[in,out]  ErrorCount        Counter used to keep track of error codes in
+                                    StatusCodeValue
+
+  @retval   EFI_SUCCESS             Successfully retrieved device ID.
+**/
+EFI_STATUS
+GetDeviceId (
+  IN      IPMI_BMC_INSTANCE_DATA  *IpmiInstance,
+  IN      EFI_STATUS_CODE_VALUE   StatusCodeValue[],
+  IN OUT  UINT8                   *ErrorCount
+  )
+{
+  EFI_STATUS                 Status;
+  UINT32                     DataSize;
+  SM_CTRL_INFO               *pBmcInfo;
+  IPMI_MSG_GET_BMC_EXEC_RSP  *pBmcExecContext;
+  UINT32                     Retries;
+
+  //
+  // Set up a loop to retry for up to PcdIpmiBmcReadyDelayTimer seconds. Calculate retries not timeout
+  // so that in case KCS is not enabled and IpmiSendCommand() returns
+  // immediately we will not wait all the PcdIpmiBmcReadyDelayTimer seconds.
+  //
+  Retries = PcdGet8 (PcdIpmiBmcReadyDelayTimer);
+  //
+  // Get the device ID information for the BMC.
+  //
+  DataSize = sizeof (IpmiInstance->TempData);
+  DEBUG ((DEBUG_INFO, "[IPMI] Getting BMC Device ID. DataSize: 0x%x Retries: %d\n", DataSize, Retries));
+  while (EFI_ERROR (
+           Status = IpmiSendCommand (
+                      &IpmiInstance->IpmiTransport,
+                      IPMI_NETFN_APP,
+                      0,
+                      IPMI_APP_GET_DEVICE_ID,
+                      NULL,
+                      0,
+                      IpmiInstance->TempData,
+                      &DataSize
+                      )
+           )
+         )

```

Comment:
nit: This could perhaps be a do while instead of a long function call in the loop condition?